File: important/PathSumII.java
// Backtracking, Tree, Depth-First Search, Binary Tree
// Amazon 5 Bloomberg 3 DoorDash 3 Zillow 2
// https://leetcode.com/problems/path-sum-ii/

// Given the root of a binary tree and an integer targetSum, return all root-to-leaf paths where the sum of the node values in the path equals targetSum. Each path should be returned as a list of the node values, not node references.

// A root-to-leaf path is a path starting from the root and ending at any leaf node. A leaf is a node with no children.

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class PathSumII {

  // bfs will cost too much space because each node will need to maintain a list
  // as a result, dfs is better for this question
  public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
    List<List<Integer>> result = new ArrayList<>();
    List<Integer> list = new ArrayList<>(); // traverse state in dfs
    this.searchSum(root, targetSum, list, result);

    return result;
  }

  // dfs
  private void searchSum(
    TreeNode root,
    int targetSum,
    List<Integer> list,
    List<List<Integer>> result
  ) {
    if (root == null) {
      return;
    }

    list.add(root.val);

    if (this.isLeaf(root) && targetSum == root.val) {
      result.add(new ArrayList<>(list));
    }
    int nextTargetSum = targetSum - root.val;
    if (root.left != null) {
      this.searchSum(root.left, nextTargetSum, list, result);
    }
    if (root.right != null) {
      this.searchSum(root.right, nextTargetSum, list, result);
    }

    list.remove(list.size() - 1);
  }

  private boolean isLeaf(TreeNode node) {
    return node.left == null && node.right == null;
  }
}

File: old/P1/PathSumII.py
# Backtracking, Tree, Depth-First Search, Binary Tree
# Amazon 5 Bloomberg 3 DoorDash 3 Zillow 2
# https://leetcode.com/problems/path-sum-ii/

# Given the root of a binary tree and an integer targetSum, return all root-to-leaf paths where the sum of
# the node values in the path equals targetSum. Each path should be returned as a list of the node values,
# not node references.

# A root-to-leaf path is a path starting from the root and ending at any leaf node. A leaf is a node with no children.


import copy
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# bfs will cost too much space because each node will need to maintain a list
# as a result, dfs is better for this question
class PathSumII:
    def pathSum(self, root: Optional[TreeNode], target: int) -> List[List[int]]:
        results = list()
        if not root:
            return results

        result_state = list()
        self.build_results(root, target, result_state, results)

        return results

    def build_results(
        self,
        root: Optional[TreeNode],
        target_remain: int,
        result_state: List[int],
        results: List[List[int]],
    ) -> None:
        if not root:
            return

        # if target_remain < root.val: return # not work for negative val

        result_state.append(root.val)

        if target_remain == root.val and self.is_leaf(root):
            results.append(copy.deepcopy(result_state))

        if root.left:
            self.build_results(
                root.left, target_remain - root.val, result_state, results
            )
        if root.right:
            self.build_results(
                root.right, target_remain - root.val, result_state, results
            )

        result_state.pop()

    def is_leaf(self, node: TreeNode) -> bool:
        return not node.left and not node.right
